### PR TITLE
Refactor representations as a module

### PR DESCRIPTION
**Pull request by mhvk**
This is a reimplementation of #13688, redoing the refactoring to ensure that (1)  there are no changes other than what is strictly necessary, and (2) any commits since that time are included (I restarted when I noticed #13688 had an import of a deprecated function that had been removed, and worried there were other changes). I avoided introducing any `__all__` in the submodules, for now just relying on importing what is relevant in the new `__init__.py`. I cherry-picked commits from #13688 that fix errors in the documentation and in serialization because of the class name changes (making a few small changes in the process). Like in #13688, I also keep a few variables in `__init__.py` that are not exported, but are used in tests, for backward compatibility.

I mostly did this since I still think the refactoring is a very good idea, and it would be helpful for #14763, which introduces more general geodetic representations, if there is a place to put them. 

Ping @nstarman who did #13688 and @eerovaher who had requested changes to it, mostly to ensure we make no actual code changes (which I think I ensured!); if we all still think this is a good idea, we should probably get it in sooner rather than later... (thankfully, `representations` is quite mature, so I don't foresee problems with backporting to 5.3).

p.s. The commits are meant to be fairly logical: for all `pytest -P coordinates` worked (except a few names in doctests after `spherical` is factored out).

EDIT: Close https://github.com/astropy/astropy/pull/13688

### CODE REVIEW

File: astropy/coordinates/representation/__init__.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    UnitSphericalCosLatDifferential,
+    RadialDifferential,
+    PhysicsSphericalDifferential,
+)
+
+# The following imports are included for backwards compatibility.
+# isort: split
+from .base import DIFFERENTIAL_CLASSES
+from .base import DUPLICATE_REPRESENTATIONS
+from .base import REPRESENTATION_CLASSES
```
Comment (nstarman) [inline astropy/coordinates/representation/__init__.py:28]: Would it be worthwhile to make these read-only?

```suggestion
from types import MappingProxyType

from .base import MappingProxyType(DIFFERENTIAL_CLASSES)
from .base import MappingProxyType(DUPLICATE_REPRESENTATIONS)
from .base import MappingProxyType(REPRESENTATION_CLASSES)
```

Or to deprecate their access from here?

**Review by nstarman — COMMENTED** [review]
This looks good and was very easy to review, stepping though the commits!

File: astropy/coordinates/representation/__init__.py (hunk 1 above)
Comment (mhvk) [reply astropy/coordinates/representation/__init__.py:28]: I agree it should be deprecated, which is relatively easy to do with `__getattr__`, but think it is best to do as a separate PR after this, so one would not be distracted by all the other changes.

File: astropy/table/serialize.py [hunk 2]
```diff
@@ -40,6 +40,23 @@
...
+    "astropy.coordinates.representation.spherical.PhysicsSphericalRepresentation",
+    "astropy.coordinates.representation.cylindrical.CylindricalRepresentation",
+    "astropy.coordinates.representation.cartesian.CartesianDifferential",
+    "astropy.coordinates.representation.spherical.UnitSphericalDifferential",
+    "astropy.coordinates.representation.spherical.SphericalDifferential",
+    "astropy.coordinates.representation.spherical.UnitSphericalCosLatDifferential",
+    "astropy.coordinates.representation.spherical.SphericalCosLatDifferential",
+    "astropy.coordinates.representation.spherical.RadialDifferential",
+    "astropy.coordinates.representation.spherical.PhysicsSphericalDifferential",
+    "astropy.coordinates.representation.cylindrical.CylindricalDifferential",
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/table/serialize.py:58]: You already opened #13689 for that purpose - where you made the suggestion of using `entrypoints`, which seems just right for this.

**Comment by mhvk** [comment]
@nstarman - I think all your points are good ones - but also feel best for this PR is to make only the most minimal changes...

**Comment by mhvk** [comment]
OK, will merge now!

**Comment by pllim** [comment]
`+3,728 −3,641` 😱

**Comment by mhvk** [comment]
Of course, mostly exactly the same lines. And a file with 3600 lines of code really is overly long!

**Pull request merged**